File: nanopay/src/net/nanopay/security/csp/CSPFilter.java
package net.nanopay.security.csp;

import java.io.IOException;
import javax.servlet.*;
import javax.servlet.http.*;

public class CSPFilter
  implements Filter {

  protected FilterConfig config_;
  protected String csp_;

  public void init(FilterConfig filterConfig) throws ServletException {
    this.config_ = filterConfig;
    csp_ = config_.getInitParameter("CONTENT_SECURITY_POLICY");
  }

  @Override
  public void doFilter(ServletRequest request, ServletResponse response,
                       FilterChain chain) throws IOException, ServletException {
    HttpServletResponse httpResponse = (HttpServletResponse) response;
    httpResponse.setHeader("Content-Security-Policy", csp_);

    chain.doFilter(request, response);
  }

  @Override
  public void destroy(){}
}

File: src/foam/nanos/security/csp/CSPFilter.js
/**
 * @license
 * Copyright 2023 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.nanos.security.csp',
  name: 'CSPFilter',

  documentation: 'Content-Security-Policy servlet filter',

  javaImplements: [ 'javax.servlet.Filter' ],

  javaImports: [
    'javax.servlet.*',
    'javax.servlet.http.HttpServletResponse'
  ],

  javaCode: `
    protected FilterConfig config_;
    protected String csp_;
  `,

  methods: [
    {
      name: 'init',
      args: 'FilterConfig config',
      javaThrows: [ 'ServletException' ],
      javaCode: `
        this.config_ = config;
        this.csp_    = config.getInitParameter("CONTENT_SECURITY_POLICY");
      `
    },
    {
      name: 'doFilter',
      args: 'ServletRequest request, ServletResponse response, FilterChain chain',
      javaThrows: [ 'java.io.IOException', 'ServletException' ],
      javaCode: `
        var httpResponse = (HttpServletResponse) response;
        httpResponse.setHeader("Content-Security-Policy", csp_);

        chain.doFilter(request, response);
      `
    },
    {
      name: 'destroy',
      javaCode: '// noop'
    }
  ]
});
